src/mongo_odm/nvp.hpp

```diff
@@ -74,17 +159,593 @@ struct Nvp {
+};
+
+/**
+ * Represents a field that does not have a name, i.e. an array element.
+ * This is used in $elemMatch expressions on scalar arrays, where the elements can be compared but
+ * don't have fields of their own.
+ * @tparam T    The type of the field.
+ */
+template <typename T>
+class FreeNvp : public NvpCRTP<FreeNvp<T>, T> {
+   public:
+    // In case this corresponds to an optional type, store the underlying type.
+    using no_opt_type = remove_optional_t<T>;
+    using type = T;
+
+    constexpr UpdateExpr<FreeNvp<T>, no_opt_type> operator=(const no_opt_type& val) const {
+        return {*this, val, "$set"};
+    }
+
+    // TODO This shouldn't have a name, but it needs to work with Expressions when building queries.
+    std::string get_name() const {
+        return "";
+    }
+};
+
+/**
+ * A CRTP base class that contains member functions for name-value pairs.
+ * These functions are identical between Nvp<...> and NvpChild<...>, but their return types are
+ * templated on the Nvp's types, so they are defined here using CRTP.
+ * @tparam NvpT The type of the name-value pair
+ * @tparam T    The type of the field referred to by the name-value pair.
+ */
+template <typename NvpT, typename T>
+class NvpCRTP {
+   public:
+    // In case this field is wrapped in an optional, store the underlying type.
+    using no_opt_type = remove_optional_t<T>;
+
+    // Type trait that checks if the given iterable class contains the same value type as either
+    // a) this Nvp's type, or
+    // b) this Nvp's value type, if this Nvp is also an iterable.
+    template <typename Iterable, typename Default = void>
+    using enable_if_matching_iterable_t = typename std::enable_if<
+        is_iterable_not_string<Iterable>::value &&
+            std::is_convertible<iterable_value<Iterable>, iterable_value<no_opt_type>>::value,
+        Default>::type;
+
+    /**
+     * Chains two name-value pairs to access a sub-field, i.e. a field with the name "parent.child".
+     * @tparam U    The type of the child NVP
+     * @param child An NVP that corresponds to a sub-field of this NVP. Its base class must be the
+     * same as this field's current type.
+     * @return      An NVP with the same base class and type as the subfield, but with a link to a
+     * parent so that its name is qualified in dot notation.
+     */
+    template <typename U>
+    constexpr NvpChild<T, U, NvpT> operator->*(const Nvp<T, U>& child) const {
+        return {child.t, child.name, *static_cast<const NvpT*>(this)};
+    }
+
+    /**
+     * Creates a sort expression that sorts documents by this field.
+     * @param ascending Whether to sort by ascending order (+1 in MongoDB syntax).
+     *                  If false, sorts by descending order (-1 in MongoDB syntax).
+     * @return a SortExpr that reprsents the sort expression {field: +/-1}.
+     */
+    constexpr SortExpr<NvpT> sort(bool ascending) const {
+        return {*static_cast<const NvpT*>(this), ascending};
```

Comment:
Using this pointer causes tests to fail, with what looks like memory alignment issues. I get errors like: 
```
/Users/raphaelkargon/Documents/mongo-cxx-odm/src/mongo_odm/test/query_builder.cpp:122: FAILED:
  REQUIRE( nvp_child.parent.parent.t == &BarParent::b )
with expansion:
  0x000000010fa78e4f == 0x0000000000000000
```
@hanumantmk thoughts?
Matt got the `self` pointer to work in a simplified example, but it fails here.